### PR TITLE
Fix bugs for router on handling response from pipeline microservices

### PR DESCRIPTION
**Pull request by zhlsunshine**
## Description

There will be timeout for router to handle response from pipeline microservices due to it need to wait for the whole response.
Using bufio to handle the response and make the new goroutine to call microservices in `mcGraphHandler`

## Issues

https://github.com/opea-project/GenAIInfra/issues/148

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)

## Dependencies

`n/a`.

## Tests

Below is the testing snapshot:

image

### CODE REVIEW

File: microservices-connector/cmd/router/main.go [hunk 1]
```diff
@@ -365,22 +367,54 @@ func routeStep(nodeName string, graph mcv1alpha3.GMConnector, input []byte, head
...
+				return
+			}
 		}
+	}()
+
+	select {
+	case <-ctx.Done():
+		log.Error(errors.New("failed to process request"), "request timed out")
+		http.Error(w, "request timed out", http.StatusGatewayTimeout)
+	case <-done:
```
Comment (gyohuangxin) [inline microservices-connector/cmd/router/main.go:417]: Could you also add some logs to suggest that the process is successful?

Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:417]: okay, make sense.

**Review by gyohuangxin — APPROVED** [review]
I've verified it worked on my environment.

File: microservices-connector/cmd/router/main.go [hunk 2]
```diff
@@ -365,22 +367,55 @@ func routeStep(nodeName string, graph mcv1alpha3.GMConnector, input []byte, head
...
-			log.Error(err, "failed to write mcGraphHandler response")
+
+		writer := bufio.NewWriter(w)
+		defer func() {
+			if err := writer.Flush(); err != nil {
+				log.Error(err, "error flushing writer when processing response")
+			}
+		}()
+
+		for start := 0; start < len(response); start += 1024 {
```
Comment (mkbhanda) [inline microservices-connector/cmd/router/main.go:401]: what is this magic number 1024 and incrementing by it?

File: microservices-connector/cmd/router/main.go [hunk 3]
```diff
@@ -365,22 +367,55 @@ func routeStep(nodeName string, graph mcv1alpha3.GMConnector, input []byte, head
...
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(statusCode)
+			if _, err := w.Write(prepareErrorResponse(err, "Failed to process request")); err != nil {
+				log.Error(err, "failed to write mcGraphHandler response")
+			}
+			return
 		}
-	} else {
 		if json.Valid(response) {
 			w.Header().Set("Content-Type", "application/json")
```
Comment (mkbhanda) [inline microservices-connector/cmd/router/main.go:390]: if json invalid what do we set in content-type? blank/nothing? is that valid?

Comment (mkbhanda) [reply microservices-connector/cmd/router/main.go:390]: There seem to be several types looking at https://stackoverflow.com/questions/23714383/what-are-all-the-possible-values-for-http-content-type-header  and we might encounter these when using genAI for image, audio etc construction. Perhaps add a note "Other types not validated at this time" type comment.

**Review by mkbhanda — COMMENTED** [review]
Please check inline comments. Happy to approve as soon as I hear from you.

File: microservices-connector/cmd/router/main.go (hunk 2 above)
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:401]: Hi @mkbhanda, the router service will write the response message back, and the router need to wait for the whole response from microservices. And this enhancement is that router service will write back as soon as there are some feedback from microservices. So I use 1024 as the length of the message to send it out.

File: microservices-connector/cmd/router/main.go (hunk 3 above)
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:390]: Hi @mkbhanda, it's a great question, router service as an entry endpoint, both the requests and responses of microservices are based on json format, so it's oaky for now. And I have some tests that the image can also be represented as base64 encode data.
Of course, there would be other changes when some other examples imported. Thanks for raising this question!

File: microservices-connector/cmd/router/main.go (hunk 2 above)
Comment (mkbhanda) [reply microservices-connector/cmd/router/main.go:401]: Just curious why you choose 1024, 1K. What are the repercussions if you had chosen 256 or 512 or 1 MB? Would it make sense to make this a constant or parameter, when dealing with different kinds of media and communication protocols for inter-microservice communication (like DMA, shared memory, UDP, TCP). After each such unit you are checking for errors.

**Review by mkbhanda — APPROVED** [review]
Thank you for the answers!

**Review by mkbhanda — APPROVED** [review]
LGTM, Thank you for introducing a constant, ChunkSize.

**Pull request merged**